src/samples/audio/smart_amp_test.c

```diff
@@ -86,6 +105,213 @@ static struct comp_dev *smart_amp_new(const struct comp_driver *drv,
 	return NULL;
 }
 
+#if CONFIG_IPC_MAJOR_4
+static void smart_amp_set_params(struct comp_dev *dev,
+				 struct sof_ipc_stream_params *params)
+{
+	struct smart_amp_data *sad = comp_get_drvdata(dev);
+	struct comp_buffer *sink;
+	struct comp_buffer __sparse_cache *sink_c;
+
+	comp_dbg(dev, "smart_amp_set_params()");
+
+	memset(params, 0, sizeof(*params));
+	params->channels = sad->config.base.audio_fmt.channels_count;
+	params->rate = sad->config.base.audio_fmt.sampling_frequency;
+	params->sample_container_bytes = sad->config.base.audio_fmt.depth / 8;
+	params->sample_valid_bytes =
+		sad->config.base.audio_fmt.valid_bit_depth / 8;
+	params->buffer_fmt = sad->config.base.audio_fmt.interleaving_style;
+	params->buffer.size = sad->config.base.ibs;
+
+	/* update sink format */
+	if (!list_is_empty(&dev->bsink_list)) {
+		struct ipc4_output_pin_format *sink_fmt = &sad->config.output_pin_fmt;
+		struct ipc4_audio_format out_fmt = sink_fmt->audio_fmt;
+
+		sink = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
+		sink_c = buffer_acquire(sink);
+		sink_c->stream.channels = out_fmt.channels_count;
+		sink_c->stream.rate = out_fmt.sampling_frequency;
+
+		audio_stream_fmt_conversion(out_fmt.depth,
+					    out_fmt.valid_bit_depth,
+					    &sink_c->stream.frame_fmt,
+					    &sink_c->stream.valid_sample_fmt,
+					    out_fmt.s_type);
+
+		sink_c->buffer_fmt = out_fmt.interleaving_style;
+		params->frame_fmt = sink_c->stream.frame_fmt;
+
+		sink_c->hw_params_configured = true;
+
+		buffer_release(sink_c);
+	}
+}
+
+static inline int smart_amp_set_config(struct comp_dev *dev, const char *data,
+				       uint32_t data_size)
+{
+	struct smart_amp_data *sad = comp_get_drvdata(dev);
+	struct sof_smart_amp_config *cfg;
+	uint32_t cfg_size;
+
+	cfg = (struct sof_smart_amp_config *)data;
+	cfg_size = data_size;
+
+	if (cfg_size != sizeof(struct sof_smart_amp_config)) {
+		comp_err(dev, "smart_amp_set_config(): invalid config size %u, expect %u",
+			 cfg_size, sizeof(struct sof_smart_amp_config));
+		return -EINVAL;
+	}
+
+	comp_dbg(dev, "smart_amp_set_config(): config size = %u", cfg_size);
+
+	memcpy_s(&sad->config, sizeof(struct sof_smart_amp_config), cfg,
+		 sizeof(struct sof_smart_amp_config));
+
+	return 0;
+}
+
+static inline int smart_amp_get_config(struct comp_dev *dev, char *data,
+				       uint32_t *data_size)
+{
+	struct smart_amp_data *sad = comp_get_drvdata(dev);
+	uint32_t cfg_size;
+	int ret;
+
+	cfg_size = sizeof(struct smart_amp_data);
+
+	if (cfg_size > *data_size) {
+		comp_err(dev, "smart_amp_get_config(): wrong config size %d",
+			 *data_size);
+	}
+
+	*data_size = cfg_size;
+
+	ret = memcpy_s(data, cfg_size, &sad->config, cfg_size);
+	assert(!ret);
```

Comment:
i know this is an irrecoverable failure but can we please limit the use of asserts and return error? simplyy ``` return memcpy_s(...)"

Reply:
sure, thanks